### ISSUE TITLE
Let games have a manifest to choose how to launch

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/itch#92]
Still have to settle on a format

**Comment by phrounz** [comment]
What I'd like to see also is a

**Comment by phrounz** [comment]
Also what would be nice: a user choice popup/window/screen/whatever to ask which executable you want to launch if there is no manifest and there are several executables (matching your operating system) in the folder.

(This is linked to issue https://github.com/itchio/itch/issues/77)

For example if the directory contains:
- working_dir
  - mygame_windowed.exe
  - mygame_fullscreen.exe

Then:

> Select which executable you want to launch:
> - mygame_windowed
> - mygame_fullscreen

For another practical example, my [Ludum Dare 34](http://ludumdare.com/compo/ludum-dare-34/?action=preview&uid=17202) game has two executables, using different rendering libraries and compiled with different environments. It took one of the two, not even sure which:
- App_VS2008_OpenGL.exe
- App_VS2013_DX_Desktop.exe

**Comment by leafo** [comment]
I like the idea of choosing the executable if there's no metadata for execution. Also the developers can already include install instructions on the website, we should display them in the app

**Comment by fasterthanlime** [comment]
> I like the idea of choosing the executable if there's no metadata for execution. Also the developers can already include install instructions on the website, we should display them in the app

Me too, just haven't had time to implement it yet. Ideally the app would be smart enough to recommend the best executable _yet_ let you choose to launch another if you want to.

**Comment by FrozenCow** [comment]
How about introducing some list of actions. That way a developer can define the different ways you can run a game. An action would have a type (like `browser`, `love`, `java` or `native`), one or more arguments (`path`, `url`, `love_version`, etc) and some way to determine whether the platform is supported (`architecture` (i686, x86_64), `platform` (windows, osx, linux), `browser` (chrome, firefox)).

If only one action matches, you could automatically start. If more match, the user should choose. It would also allow to _eventually_ use Itch on something like Android.

**Comment by TangentFoxy** [comment]
- Also, if there is no executable in the root directory of a zip, but there is a folder in it, check that folder for executables. I see a lot of games package as `zip->folder->exec`.
- The user should be able to add a custom launch command for a game (or multiple).

**Comment by fasterthanlime** [comment]
> Also, if there is no executable in the root directory of a zip, but there is a folder in it, check that folder for executables. I see a lot of games package as zip->folder->exec.

the app handles that, it'll look for execs nested arbitrarily deep. shortest path with best score wins, cf #86

**Comment by TangentFoxy** [comment]
Cool! (What about my 2nd suggestion?)

**Comment by fasterthanlime** [comment]
> What about my 2nd suggestion?

Every 'multiple option' idea is interesting but requires non-trivial UI work, so keeping that for later. For now I just want every game to have at least 1 launch option that works by default.

**Comment by initials** [comment]
I always have two exes in my games, and most games are alphabetically after the engine, it opens the bare engine.

Basically I'm hoping this manifest can happen soon.

**Comment by initials** [comment]
I tried to solve this by having a cmd file, and it doesn't seem to scan for .cmd files or .bat files, although I renamed the file to a bat after downloading.

**Comment by fasterthanlime** [comment]
It does scan for .bat files, but it only scans once when installing the game

**Comment by Ragzouken** [comment]
as i said on twitter, it'd be cool to be able to specify a pdf/txt file to provide as a manual to the user

**Comment by fasterthanlime** [comment]
So, I'm thinking of doing this sooner rather than later. 

I'm thinking of using [TOML](https://github.com/toml-lang/toml) (argue now or forever hold your peace)

``` toml
[actions.play]
path = "StartGame.bat"

[actions.editor]
path = "Tools/Editor.exe"
args = ["--enable-itch-workshop"]

[actions.manual]
path = "Docs/Manual.pdf"
```

More advanced example:

``` toml
[actions.play]
paths = ["Game32.exe", "Game64.exe"]
scopes = ["profile:me", "profile:friends", "twitter:post"]
languages = ["en", "fr", "es", "ru"]

[[actions.play.flags]]
label = "OpenGL 3.3 workaround"
flag = "-force-glcore33"
```

Tagging platforms/archs explicitly shouldn't be required, scopes are API-related (when unspecified, would be "profile:me"), flags would be implemented as checkboxes in the game's "Launch properties" dialogs (cf. #366).

I don't really believe "type" is needed: there's pretty much one clear action for most types of files you'd throw at itch.

Input welcome!

**Comment by fasterthanlime** [comment]
> Tagging platforms/archs explicitly shouldn't be required

Expanding on this: it's easy for the itch app to look at an executable and figure out if it's 32-bit or 64-bit. The part where explicit tagging would be useful is [when uploading](https://itch.io/docs/butler/pushing.html) separate 32 and 64-bit builds of a game, so the app knows what to download.

**Comment by fasterthanlime** [comment]
Started cleaning up the 'launching process' for apps. Providing API access (to /api/1/jwt/me) is a top priority for me but have to go through with research in #670 to make sure nothing about the way manifests work is incompatible (in principle) with sandboxing.

**Comment by fasterthanlime** [comment]
Shipped in v18.0.0! (dependencies will be in a separate issue)
